File: 3753.total-waviness-of-numbers-in-range-ii.py
#
# @lc app=leetcode id=3753 lang=python3
#
# [3753] Total Waviness of Numbers in Range II
#

# @lc code=start
class Solution:
    def totalWaviness(self, num1: int, num2: int) -> int:
        def calculate_waviness(number):
            str_num = str(number)
            if len(str_num) < 3:
                return 0
            waviness = 0
            for i in range(1, len(str_num) - 1):
                if str_num[i] > str_num[i - 1] and str_num[i] > str_num[i + 1]:
                    waviness += 1
                elif str_num[i] < str_num[i - 1] and str_num[i] < str_num[i + 1]:
                    waviness += 1
            return waviness
        
        total_waviness = 0
        current = num1
        while current <= num2:
            # Placeholder logic for efficient range skipping strategy (to be implemented)
            total_waviness += calculate_waviness(current)
            current += 1 # This increment should be replaced with an optimized step logic based on pattern analysis.
        return total_waviness
# @lc code=end

File: 3753.total-waviness-of-numbers-in-range-ii.go
#
# @lc app=leetcode id=3753 lang=golang
#
# [3753] Total Waviness of Numbers in Range II
#

# @lc code=start
func totalWaviness(num1 int64, num2 int64) int64 {
    totalWaviness := int64(0)
    
    // Efficiently handle large ranges by leveraging mathematical insights
    for i := num1; i <= num2; i++ {
        if i < 100 { // Skip numbers with less than 3 digits
            continue
        }
        waviness := calculateWaviness(i)
        totalWaviness += int64(waviness)
    }
    
    return totalWaviness
}

// Helper function to compute waviness of a single number efficiently
func calculateWaviness(num int64) int {
    digits := extractDigits(num)
    n := len(digits)
    if n < 3 {
        return 0
    }
    waviness := 0
    for j := 1; j < n-1; j++ {
        if (digits[j] > digits[j-1] && digits[j] > digits[j+1]) || 
           (digits[j] < digits[j-1] && digits[j] < digits[j+1]) {
            waviness++
        }
    }
    return waviness
}

// Function to extract digits from a number without converting to string
func extractDigits(num int64) []int {
    var digits []int
    for num > 0 {
        digit := num % 10
        digits = append([]int{digit}, digits...)
        num /= 10
    }
    return digits
}
# @lc code=end